File: ios/Sources/TaigiKeyboard/NextWord/AutocompleteContextBooster.swift
import Foundation

// MARK: - Shared-Core Candidate
// Pure logic, Foundation-only. Eligible for cross-platform extraction.

/// Reorders candidate words to surface those whose first character matches
/// a next-word prediction bigram. The predictions are computed elsewhere
/// (`NextWordService`) and passed in as a set of first-character strings,
/// so this type stays pure: no DB, no async, no side effects.
///
/// Moved here from `AutocompleteService.swift` to co-locate with NextWord.
enum AutocompleteContextBooster {
    /// - Parameters:
    ///   - words: 候選詞列表（已排序）
    ///   - predictedFirstChars: 由上一個選字的 bigram 預測出的首字集合
    /// - Returns: boosted 區段 + 其餘（各自保留原順序）
    static func boost(words: [TaigiWord], predictedFirstChars: Set<String>) -> [TaigiWord] {
        guard !predictedFirstChars.isEmpty else { return words }

        var boosted: [TaigiWord] = []
        var rest: [TaigiWord] = []
        for word in words {
            if let firstChar = word.displayText.first,
               predictedFirstChars.contains(String(firstChar))
            {
                boosted.append(word)
            } else {
                rest.append(word)
            }
        }
        return boosted + rest
    }
}

File: android/app/src/main/java/com/siansiansu/taigikeyboard/ime/text/composing/AutocompleteContextBooster.kt
// region Shared-Core Candidate
// Pure logic, Kotlin stdlib only. Eligible for cross-platform extraction.
// endregion
package com.siansiansu.taigikeyboard.ime.text.composing

import com.siansiansu.taigikeyboard.ime.dictionary.TaigiWord

/**
 * Reorder candidates so that entries whose display-text first character is
 * a bigram-predicted next character float to the top. The prediction set
 * is computed by the caller (via `NextWordService`); this helper stays
 * pure — no DB, no async, no side effects.
 *
 * Mirrors `ios/Sources/TaigiKeyboard/NextWord/AutocompleteContextBooster.swift`.
 */
object AutocompleteContextBooster {
    /**
     * Preserves original order within the boosted and rest partitions.
     * Empty `predictedFirstChars` short-circuits: returns `words` untouched.
     */
    fun boost(
        words: List<TaigiWord>,
        predictedFirstChars: Set<String>,
    ): List<TaigiWord> {
        if (predictedFirstChars.isEmpty()) return words

        val boosted = mutableListOf<TaigiWord>()
        val rest = mutableListOf<TaigiWord>()
        for (word in words) {
            val firstChar = word.displayText.firstOrNull()?.toString()
            if (firstChar != null && firstChar in predictedFirstChars) {
                boosted.add(word)
            } else {
                rest.add(word)
            }
        }
        return boosted + rest
    }
}
